### PR TITLE
support making loadtest users

### PR DESCRIPTION
**Pull request by czue**
http://manage.dimagi.com/default.asp?165814

image

probably best reviewed commit by commit.

dependent on https://github.com/dimagi/commcare-hq/pull/6398 for tests (these are the first 3 commits)

this is a rather limited implementation, but i think good enough for mwellcare and can be improved later.

@benrudolph has context on the feature and @snopoke probably is most familiar with the code. cc @kaapstorm 

the meat is in https://github.com/dimagi/commcare-hq/commit/484189da8c1823ef9218d0ba2a9089fd5384ced9 where i've left some comments

### CODE REVIEW

File: corehq/ex-submodules/casexml/apps/case/tests/util.py [hunk 1]
```diff
@@ -19,6 +19,30 @@
...
+
+    def get_case_id(self):
+        if self.version == V1:
+            return self.xml_element.findtext('{{{0}}}case_id'.format(get_case_xmlns(self.version)))
+        else:
+            return self.xml_element.get('case_id')
+
+    def get_case_name(self):
+        assert self.version == V2, 'get_case_name not yet supported for legacy V1 casexml'
+        # note: there has to be a better way to work with namespaced xpath.... right?!?!
```
Comment (benrudolph) [inline corehq/ex-submodules/casexml/apps/case/tests/util.py]: yea it's pretty horrific. i had to dive into a little bit when writing a script. you can see how app_manager does it? https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/app_manager/xform.py#L27. @snopoke would definitely know more

Comment (benrudolph) [inline corehq/ex-submodules/casexml/apps/case/tests/util.py]: oops don't think my comment was relevant

Comment (benrudolph) [inline corehq/ex-submodules/casexml/apps/case/tests/util.py]: although i think the approach app_manager takes is to wrap the elements so you don't have to specify the namespaces each time. could be wrong

File: corehq/ex-submodules/casexml/apps/case/tests/util.py [hunk 2]
```diff
@@ -176,7 +201,7 @@ def post_util(create=False, case_id=None, user_id=None, owner_id=None,
                       case_id=case_id,
                       user_id=user_id or uid(),
                       owner_id=owner_id or uid(),
-                      case_type=case_type or DEFAULT_TEST_TYPE,
+                      case_type=case_type or 'test',
```
Comment (benrudolph) [inline corehq/ex-submodules/casexml/apps/case/tests/util.py]: why remove the constant? won't we have to use it again to find the test cases

Comment (czue) [inline corehq/ex-submodules/casexml/apps/case/tests/util.py]: it was only used in the one place so it just seemed unnecessary

File: corehq/ex-submodules/casexml/apps/phone/restore.py [hunk 3]
```diff
@@ -342,13 +344,46 @@ def get_restore_class(user):
...
 
 
+def _get_loadtest_factor(domain, user):
+    """
+    Gets the loadtest factor for a domain and user. Is always 1 unless
+    both the toggle is enabled for the domain, and the user has a non-zero,
+    non-null factor set.
+    """
+    if domain and ENABLE_LOADTEST_USERS.enabled(domain.name):
+        return getattr(user, 'loadtest_factor', 1) or 1
```
Comment (benrudolph) [inline corehq/ex-submodules/casexml/apps/phone/restore.py]: don't think the `or 1` is necessary

**Comment by benrudolph** [comment]
sweet, way to pump this out. i'll let @snopoke take a look before merging

File: corehq/ex-submodules/casexml/apps/case/tests/util.py (hunk 2 above)
Comment (benrudolph) [inline corehq/ex-submodules/casexml/apps/case/tests/util.py]: :+1:

File: corehq/ex-submodules/casexml/apps/phone/restore.py (hunk 3 above)
Comment (czue) [inline corehq/ex-submodules/casexml/apps/phone/restore.py]: I think it is, although it's pretty paranoid.
```
>>> class foo(object): pass
... 
>>> foo.property = None
>>> getattr(foo, 'property', 1)
>>> getattr(foo, 'property', 1) or 1
1
```

**Comment by czue** [comment]
:+1: to waiting for @snopoke

File: corehq/ex-submodules/casexml/apps/phone/restore.py (hunk 3 above)
Comment (benrudolph) [inline corehq/ex-submodules/casexml/apps/phone/restore.py]: ah you're right, i saw this line: https://github.com/dimagi/commcare-hq/pull/6400/files#diff-9d9075c8ac384b5859ac5aa7120651e5R234 and thought that it was protected from being less than 1 or Falsey, but that check is only on the form so potentially you could hard code the model to have a different

File: corehq/apps/users/forms.py [hunk 4]
```diff
@@ -230,6 +231,11 @@ def direct_properties(self):
 
 
 class UpdateCommCareUserInfoForm(BaseUserInfoForm, UpdateUserRoleForm):
+    loadtest_factor = forms.IntegerField(
+        required=False, min_value=1, max_value=50000,
```
Comment (snopoke) [inline corehq/apps/users/forms.py]: 50000! wow, that's big. I woudn't have though anything more than 10000 or even 1000 was necessary.

File: corehq/ex-submodules/casexml/apps/case/tests/util.py (hunk 1 above)
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/case/tests/util.py]: you could extend `WrappedNode` which handles the namespace formatting: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/app_manager/xform.py#L125-125

File: corehq/ex-submodules/casexml/apps/case/tests/util.py [hunk 5]
```diff
@@ -19,6 +19,30 @@
 from casexml.apps.case.util import post_case_blocks
 
 
+class RestoreCaseBlock(object):
```
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/case/tests/util.py]: Nice

**Comment by snopoke** [comment]
This will only work for the first sync, after that the state hashes won't match. Is that intentional? Also the phone won't be able to submit forms for any of the exploded cases (I think the form processing will error when trying to find the case).

Looking at the spec I think these caveats are OK but should be documented somewhere (perhaps even in the help text for the loadtest factor field).

File: corehq/apps/users/forms.py (hunk 4 above)
Comment (czue) [inline corehq/apps/users/forms.py]: yeah, i was thinking that maybe someone might want to make one case and then explode it to 50000. maybe that's too aggressive though.

**Comment by czue** [comment]
@snopoke yeah - good call on better documenting. thanks for thorough review!

this would be a great use case for documentation in the feature flag framework somehow. i might explore that and see if there's a quick way to do it.

**Comment by czue** [comment]
https://github.com/dimagi/commcare-hq/pull/6408

**Pull request merged**